### ISSUE TITLE
Fragment repository implementation should be imported even it's not in basePackages of @EnableJpaRepositories

### ISSUE THREAD

**Issue by quaff** [spring-projects/spring-data-jpa#3287]
## Given:
```java
package com.example.demo;

public interface CustomRepository<T> {
	
	void customSave(T entity);

}
```

```java
package com.example.demo;

public class CustomRepositoryImpl<T> implements CustomRepository<T> {

	@Override
	public void customSave(T entity) {
		throw new UnsupportedOperationException("Not Implemented");
	}

}

```

```java
package test;

import org.springframework.data.jpa.repository.JpaRepository;

import com.example.demo.CustomRepository;

public interface TestEntityRepository extends JpaRepository<TestEntity, Long>, CustomRepository<TestEntity> {

}
```
## Then:
`com.example.demo.CustomRepositoryImpl` should be registered as repository fragment implementation implicitly since `com.example.demo.CustomRepository` is declared as repository fragment.

## Actual:
Here is a failed test
```java
package test;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.domain.EntityScan;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
import org.springframework.test.context.ContextConfiguration;

@DataJpaTest
//@EnableJpaRepositories(basePackageClasses = { TestEntityRepository.class, CustomRepository.class }) // will works
@EnableJpaRepositories(basePackageClasses = { TestEntityRepository.class })
@EntityScan(basePackageClasses = TestEntity.class)
@ContextConfiguration(classes = TestEntityRepositoryTests.class)
public class TestEntityRepositoryTests {

	@Autowired
	private TestEntityRepository testEntityRepository;

	@Test
	void test() {
		assertThatThrownBy(() -> testEntityRepository.customSave(new TestEntity()))
				.isInstanceOf(UnsupportedOperationException.class).hasMessage("Not Implemented");
	}

}
```

Minimal reproducible project: [repository-scan.zip](https://github.com/spring-projects/spring-data-jpa/files/13806871/repository-scan.zip)

**Comment by quaff** [comment]
It's a valid use case that reuse fragment repositories as common library share by applications, normally application use different base package name, then `@Enable…Repositories` must include both base package name and common library base package name, It would be great if we can use `@EnableJpaRepositories` instead of `@EnableJpaRepositories(basePackages={"com.library", "com.application"})`, the library fragment repository should be automatic activated if it's extended by application repository.

I propose that `$FragmentInterfaceName + $ImplementationPostfix` should be default implementation no matter it's in base packages or not, It could be overridden by other implementation in base packages.


Or It should be available if it's registered as spring bean:
```java
package test;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.domain.EntityScan;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.context.annotation.Bean;
import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
import org.springframework.test.context.ContextConfiguration;

import com.example.demo.CustomRepository;
import com.example.demo.CustomRepositoryImpl;

@DataJpaTest
@EnableJpaRepositories(basePackageClasses = { TestEntityRepository.class })
@EntityScan(basePackageClasses = TestEntity.class)
@ContextConfiguration(classes = TestEntityRepositoryTests.Config.class)
public class TestEntityRepositoryTests {

	@Autowired
	private TestEntityRepository testEntityRepository;

	@Test
	void test() {
		assertThatThrownBy(() -> testEntityRepository.customSave(new TestEntity()))
				.isInstanceOf(UnsupportedOperationException.class).hasMessage("Not Implemented");
	}

	static class Config {
		@Bean
		CustomRepository<?> customRepositoryImpl() {
			return new CustomRepositoryImpl<>();
		}
	}
}

```

In a nutshell, it's not a good idea that `@Enable…Repositories` include third party library package name, It should be available if it's on classpath, It's idiomatic in spring boot. WDYT? @mp911de
